### PR TITLE
Add SkipRequest field to proxy to be able to blacklist requests

### PR DESCRIPTION
**Pull request by nxvl**
- [x] Add tests
  @rainforestapp/core

### CODE REVIEW

File: mitm.go [hunk 1]
```diff
@@ -107,6 +107,8 @@ type Proxy struct {
 	// proxies it should be set to HTTPSDirector.
 	Director func(*http.Request)
 
+	SkipRequest func(*http.Request) bool
```
Comment (shosti) [inline mitm.go:114]: Could you add documentation here?

File: mitm.go [hunk 2]
```diff
@@ -178,6 +183,10 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	p.proxyMITM(sc, cc)
 }
 
+func SkipNone(req *http.Request) bool {
```
Comment (shosti) [inline mitm.go:191]: Also basic doc here? (Or you can make it private; it's just nice if all public functions have documentation.)

File: mitm_test.go [hunk 3]
```diff
@@ -393,6 +393,49 @@ func TestWebsocket(t *testing.T) {
...
+
+	go func() {
+		if err := http.Serve(l, p); err != nil {
+			if !strings.Contains(err.Error(), "use of closed network") {
+				t.Fatal("Serve:", err)
+			}
+		}
+	}()
+
+	req, _ := http.NewRequest("GET", "http://www.google.com/", nil)
```
Comment (shosti) [inline mitm_test.go]: I think this is actually calling out to google? Let's try to keep the unit tests independent. Can you set up a test server (with httptest.NewServer) instead?

Comment (nxvl) [inline mitm_test.go]: kk

File: mitm.go [hunk 4]
```diff
@@ -107,6 +107,11 @@ type Proxy struct {
 	// proxies it should be set to HTTPSDirector.
 	Director func(*http.Request)
 
+	// SkipRequest is a function used to skip some requests from being
+	// proxied. If it returns true, the request passes by without being
+	// wrapped. If false, it's wrapped and proxied.
```
Comment (shosti) [inline mitm.go]: :+1: might want to also add that the default is SkipNone.

File: mitm.go [hunk 5]
```diff
@@ -178,6 +186,11 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	p.proxyMITM(sc, cc)
 }
 
+// By default we don't skip any request and proxy all of them.
```
Comment (shosti) [inline mitm.go]: Super-nitpicky, but the convention is to always start the doc with the function name (so something like "SkipNone indicates that no requests should be skipped." or something like that).

**Comment by nxvl** [comment]
@shosti ready to go

**Pull request merged**